### ISSUE TITLE
Packaging for NixOS troubles

### ISSUE THREAD

**Issue by Ramblurr** [tchapi/davis#154]
Hi @tchapi  

I'm working on getting davis packages in the NixOS linux distro.

One unique feature of NixOS is that the directory in which packaged files find themselves is immutable. Unfortunately Symfony always tries to load the `.env` (and siblings) from the directory containing `composer.json`. On NixOS this is a read-only directory, so the user's .env config cannot be placed there.

I've developed a patch that hints Symfony to look at a different location. This allows the PHP code to reside in one directory (the place where the package manager puts it), and the user's `.env[.prod]` and `var/[cache|log]` to reside in a read-write directory (for example `/var/lib/davis`, as is the convention in NixOS).

The patch is quite minimal and simply adds a few points in the `index.php`, `bin/console`, and `src/Kernel.php` where the following environment variables can be read:

* `RUNTIME_DIRECTORY`: where .env files should be read from
* `CACHE_DIRECTORY`: where `var/cache` should be
* `LOG_DIRECTORY`: where `var/log` / `kernel.logs_dir` should be

The patch: https://github.com/Ramblurr/nixpkgs/commit/e4812f1cd9607ca4aeb6eee24b910d203fda7efa#diff-e069fa0d3dfa39bbfec0a8154d2c2b4f3c9764b319b94c8e73156eae040bc0b3

Would you be willing to accept these changes here upstream? If so I can make any changes you request and open a PR.

**Comment by tchapi** [comment]
Hi @Ramblurr 

Interesting take, thanks! 

### Cache and log dirs

Taking cache as an example but it applies for both:

```diff
+    public function getCacheDir(): string
+    {
+        if (getenv('CACHE_DIRECTORY') !== false) {
+            return getenv('CACHE_DIRECTORY') . '/' . $this->getEnvironment();
+        }
+        return parent::getCacheDir();
+    }
```

Environment files are supposed to be overridable by environment, so if we allow setting a env var for these directory, I'd rather have the full path without adding anything (the environment) to it. It's also less error-prone as you don't have to know the inner workings to guess that your log / cache will be one level deeper than expected.


### Env file

Would the `composer.json` [override](https://symfony.com/doc/current/configuration/override_dir_structure.html#override-the-environment-dotenv-files-directory) work for you?

Because you won't have access to the env file if you don't know its path, so you can't `getenv` the value from the file if you haven't loaded the `.env` file itself. This only allows to do `RUNTIME_DIRECTORY=/tmp bin/console` for instance, setting explicitly the env var in the cli, which is not what you want I think?

**Comment by Ramblurr** [comment]
Re cache and log does: yes that makes sense actually. I agree

Re composer.json override: I'm convicted on this because if I understand right, our package would have to patch the composer.json, which I'd rather avoid (I'd rather avoid any patches if possible). You can see in the link you shared that I implemented the alternative method of setting setting dotenv_path in the PHP entrypoint  files.

Ideally we could specify one envvar like RUNTIME_DIRECTORY which points to a directory that 1) cache and logs would be written to and 2) further .env.local/ .env.prod could be loaded from.
 
Does that make sense?

**Comment by tchapi** [comment]
I understand the point, what I'm saying is that it's not exactly working :)

You won't be able to run the console (bin/console) or execute the index.php file to serve the interface without explicitly specifying the env everytime, and I'm not sure how you want to achieve that? you'd have to have a specific NixOS env file on top of the Davis env file (sorry I'm not an expert in NixOS)?

**Comment by Ramblurr** [comment]
> You won't be able to run the console (bin/console) or execute the index.php file to serve the interface without explicitly specifying the env everytime, and I'm not sure how you want to achieve that? you'd have to have a specific NixOS env file on top of the Davis env file (sorry I'm not an expert in NixOS)?

Ah yes this is exactly how we are doing it right now. It's all systemd services.

We run `index.php` via a php-fpm service, and run `bin/console doctrine:migrations:migrate`  via a one-shot systemd service. 

Both systemd services are run with:

```
Environment=RUNTIME_DIRECTORY="${cfg.dataDir}";
Environment=CACHE_DIRECTORY="${cfg.dataDir}/var/cache";
Environment=LOG_DIRECTORY="${cfg.dataDir}/var/log";
```

(where cfg.dataDir is something like /var/lib/davis)

Reducing that to just `RUNTIME_DIRECTORY` is a good suggestion. It's ok to have to specify at least one env var.

Then we copy the root default `.env` to `/var/lib/davis` and write the rest of the user's config to a `.env.local` file in `/var/lib/davis` and then Symfony's dotenv picks it up (when my patches are applied).

For the user to be able to manually call the console (for various maintenance tasks or troubleshooting), we also provide a simple wrapper script in `/var/lib/davis/console` that wraps the actual `bin/console` with the right env var.

**Comment by tchapi** [comment]
Ok, this makes sense.

So what I would suggest:
- one PR for the cache and log, leveraging the .env file. A warning on that: we already have `LOG_FILE_PATH` in the env, so we should give precedence to it since some users use it to output the logs to a stream for instance
- one PR for the `RUNTIME_DIRECTORY` (I would maybe favor `ENV_DIRECTORY`?)

How does that sound?

**Comment by Ramblurr** [comment]
Sorry, was offline over the weekend.

I'll open some PRs today.

**Comment by Ramblurr** [comment]
PRs are open.

**Comment by Ramblurr** [comment]
Fixed with #156  thanks for collaboration @tchapi !

**Comment by Ramblurr** [comment]
Yup! Its working it's way through PR reviews, I'll post it once it's usable and ready. 

Basically I'm waiting for a new release so I don't have to ship my patch :)
